sdks/java/io/jdbc/src/main/java/org/apache/beam/sdk/io/jdbc/JdbcIO.java

```diff
@@ -1353,14 +1236,136 @@ void set(
 
     @Override
     public PCollection<Void> expand(PCollection<T> input) {
-      checkArgument(getStatement() != null, "withStatement() is required");
+      WriteVoid<T> spec = this;
       checkArgument(
-          getPreparedStatementSetter() != null, "withPreparedStatementSetter() is required");
-      checkArgument(
-          (getDataSourceProviderFn() != null),
+          (spec.getDataSourceProviderFn() != null),
           "withDataSourceConfiguration() or withDataSourceProviderFn() is required");
+      // fixme: validate invalid table input
+      if (input.hasSchema() && !spec.hasStatementAndSetter()) {
```

Comment:
I expect that this large part of code below is mostly copied from `Write`. Is it correct?

Reply:
Yes, you're right. I moved this part of the code because it will let to use the class without the statement and at the same time it just extends this class without breaking compatibility.